src/markov/markov_approx.jl

```diff
@@ -219,3 +223,567 @@ function estimate_mc_discrete{T}(X::Vector{T})
+        q = construct_prior_guess(condMean[:, ii], Nm, y1D, method)
+
+        # Make sure all elements of the prior are stricly positive
+        q[q.<kappa] = kappa
+
+        for jj = 1:M
+            # Try to use intelligent initial guesses
+            if ii == 1
+                lambdaGuess = zeros(2)
+            else
+                lambdaGuess = lambdaBar[(jj-1)*2+1:jj*2, ii-1]
+            end
+
+            # Maximum entropy optimization
+            if nMoments == 1 # match only 1 moment
+                temp[jj, :], _, _ = discrete_approximation(y1D[jj, :],
+                    X -> (reshape(X, 1, Nm)-condMean[jj, ii])/scalingFactor[jj],
+                    [0.0], reshape(q[jj, :], 1, Nm), [0.0])
+            else # match 2 moments first
+                p, lambda, momentError = discrete_approximation(y1D[jj, :],
+                    X -> polynomial_moment(X, condMean[jj, ii], scalingFactor[jj], 2),
+                    [0; 1]./(scalingFactor[jj].^(1:2)),
+                    reshape(q[jj, :], 1, Nm), lambdaGuess)
+                if norm(momentError) > 1e-5 # if 2 moments fail, just match 1 moment
+                    warn("Failed to match first 2 moments. Just matching 1.")
+                    temp[jj, :], _, _ = discrete_approximation(y1D[jj, :],
+                        X -> (X-condMean[jj,ii])/scalingFactor[jj],
+                        0, reshape(q[jj, :], 1, Nm), 0)
+                    lambdaBar[(jj-1)*2+1:jj*2, ii] = zeros(2,1)
+                elseif nMoments == 2
+                    lambdaBar[(jj-1)*2+1:jj*2, ii] = lambda
+                    temp[jj, :] = p
+                else # solve maximum entropy problem sequentially from low order moments
+                    lambdaBar[(jj-1)*2+1:jj*2, ii] = lambda
+                    for mm = 4:2:nMoments
+                        lambdaGuess = vcat(lambda, 0.0, 0.0) # add 0 to previous lambda
+                        pnew, lambda, momentError = discrete_approximation(y1D[jj,:],
+                            X -> polynomial_moment(X, condMean[jj,ii],
+                                                    scalingFactor[jj], mm),
+                            gaussianMoment[1:mm]./(scalingFactor[jj].^(1:mm)),
+                            reshape(q[jj, :], 1, Nm), lambdaGuess)
+                        if !(norm(momentError) < 1e-5)
+                            warn(
+                            "Failed to match first $mm moments.  Just matching $(mm-2).")
+                            break
+                        else
+                            p = pnew
+                        end
+                    end
+                    temp[jj, :] = p
+                end
+            end
+        end
+        P[ii, :] .= vec(prod(allcomb3(temp'), 2))
+    end
+
+    X = C*D + repmat(mu, 1, Nm^M) # map grids back to original space
+
+    M != 1 || (return MarkovChain(P, vec(X)))
+    return MarkovChain(P, [X[:, i] for i in 1:Nm^M])
```

Comment:
I also like the idea of that, but the downside is that `Vector{Tuple{M,T}}` is not type stable (`M` is not a compile time value).

There are some tricks we can do to make that happen, but it might not be worth it...